src/main/clojure/clara/rules/engine.cljc

```diff
@@ -389,17 +422,96 @@
+       (keep child-type)
+       (map :name)
+       (distinct)
+       (sort)))
+
+(defn- list-of-names
+  "Returns formatted string with correctly pluralized header and
+  list of names. Returns nil if no such node is found."
+  [singular plural prefix names]
+  (let [msg-for-unnamed (str "  An unnamed " singular ", provide names to your "
+                             plural " if you want them to be identified here.")
+        names-string (->> names
+                          (sort)
+                          (map #(if (nil? %) msg-for-unnamed %))
+                          (map #(str prefix "  " %))
+                          (string/join "\n"))]
+    (if (pos? (count names))
+      (str prefix plural ":\n" names-string "\n"))))
+
+
+(defn- single-condition-message
+  [condition-number [condition-definition terminals]]
+  (let [productions (->> terminals
+                         (filter (comp #{:production} first))
+                         (map second))
+        queries (->> terminals
+                     (filter (comp #{:query} first))
+                     (map second))
+        production-section (list-of-names "rule" "rules" "   " productions)
+        query-section (list-of-names "query" "queries" "   " queries)]
+    (string/join
+     [(str (inc condition-number) ". " condition-definition "\n")
+      production-section
+      query-section])))
+
+(defn- throw-constraint-exception
+  "Adds a useful error message when executing a constraint node raises an exception."
+  [error node fact env]
+  (let [exp (:constraint-exp error)
+        bindings (:bindings error)
+        bindings-description (if (empty? bindings)
+                               "with no bindings\n"
+                               (str "with bindings\n  " bindings))
+        message-header (string/join ["Constraint exception.\n"
+                                     "Exception raised during execution of constraint\n"
+                                     (str "  " exp "\n")
+                                     "when processing fact\n"
+                                     (str "  " (pr-str fact) "\n")
+                                     (str bindings-description "\n")
+                                     "Conditions:\n"])
+        conditions-and-rules (get-conditions-and-rule-names node)
+        condition-messages (->> conditions-and-rules
+                                (map-indexed single-condition-message)
+                                (string/join "\n"))
+        cause (:exception error)
+        message (str message-header "\n" condition-messages)
+        info (-> error
+                 (dissoc :exception)
+                 (merge {:conditions-and-rules conditions-and-rules
+                         :fact fact
+                         :node node}))]
+    (throw (ex-info message info cause))))
+
+(defn- alpha-node-matches
+  [facts env activation node]
+  (for [fact facts
+        :let [[bindings error] (activation fact env)]
```

Comment:
Similar to the above comments...rather than returning a [bindings error] tuple, I wonder if we could just throw a specific exception we'd only catch for this case that contains the error, saving us the creation/destructuring of a tuple that only helps in an exception case.